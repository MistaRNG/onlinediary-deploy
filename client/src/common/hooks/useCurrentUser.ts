import { useDispatch, useSelector } from "react-redux";
import { useEffect } from "react";
import { getCurrentUser } from "features/auth/authSlice";
import { RootState, AppDispatch } from "app/store";

/**
 * Custom hook that manages the current user's state by checking and fetching the username.
 * 
 * This hook uses Redux to check if the username has been retrieved. If not, it triggers
 * a dispatch to fetch the current user information. The hook returns the status of the
 * username check and the username itself.
 *
 * @returns {Object} - An object containing:
 * - `hasCheckedUsername`: A boolean indicating if the username has been checked.
 * - `username`: The username of the current user, or null if not available.
 */
const useCurrentUser = () => {
  const dispatch: AppDispatch = useDispatch();

  const { hasCheckedUsername, username } = useSelector((state: RootState) => {
    const username = typeof state.username === 'string' ? state.username : null;
    return { hasCheckedUsername: username !== null, username };
  });

  useEffect(() => {
    if (!hasCheckedUsername) {
      dispatch(getCurrentUser());
    }
  }, [hasCheckedUsername, dispatch]);

  return { hasCheckedUsername, username };
};

export default useCurrentUser;