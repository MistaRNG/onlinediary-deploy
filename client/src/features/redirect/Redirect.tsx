import { useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { getTodayDate } from "common/helpers";
import Loading from "../loading/Loading";
import useCurrentUser from "common/hooks/useCurrentUser";

const Redirect: React.FC = () => {
  const navigate = useNavigate();
  const { username } = useCurrentUser();

  useEffect(() => {
    if (username) {
      const url = `/journal/${getTodayDate()}`;
      navigate(url);
    } else {
      navigate("/login");
    }
  }, [username, navigate]);

  return <Loading />;
};

export default Redirect;
