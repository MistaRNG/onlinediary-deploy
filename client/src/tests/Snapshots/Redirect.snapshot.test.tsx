import React from 'react';
import { render } from '@testing-library/react';
import Redirect from '../../features/redirect/Redirect';
import { useNavigate } from 'react-router-dom';
import useCurrentUser from '../../common/hooks/useCurrentUser';
import '@testing-library/jest-dom';
import { describe, it, jest, beforeEach, expect } from '@jest/globals';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

jest.mock('../../common/helpers', () => ({
  getTodayDate: jest.fn(() => '2024-09-06'),
}));

jest.mock('../../common/hooks/useCurrentUser', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../features/loading/Loading', () => () => <div>Loading...</div>);

describe('Redirect Component', () => {
  let mockNavigate: jest.Mock;

  beforeEach(() => {
    mockNavigate = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
  });

  it('should navigate to the journal page when the user is logged in', () => {
    // ARRANGE
    (useCurrentUser as jest.Mock).mockReturnValue({ username: 'testuser' });

    // ACT
    render(<Redirect />);

    // ASSERT
    expect(mockNavigate).toHaveBeenCalledWith('/journal/2024-09-06');
  });

  it('should navigate to the login page when the user is not logged in', () => {
    // ARRANGE
    (useCurrentUser as jest.Mock).mockReturnValue({ username: null });

    // ACT
    render(<Redirect />);

    // ASSERT
    expect(mockNavigate).toHaveBeenCalledWith('/login');
  });

  it('matches the snapshot', () => {
    // ARRANGE
    (useCurrentUser as jest.Mock).mockReturnValue({ username: null });

    // ACT
    const { asFragment } = render(<Redirect />);

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });
});
