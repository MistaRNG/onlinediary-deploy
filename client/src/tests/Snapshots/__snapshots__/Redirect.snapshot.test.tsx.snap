// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redirect Component matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
